drivers/sensor/paw3212/paw3212.c

```diff
@@ -0,0 +1,609 @@
+	 * depend on automatic integer promotion (it will convert s16_t to int).
+	 * To expand sign we cast s16 to unsigned int and left shift it then
+	 * cast it to signed integer and do the right shift. Since type is
+	 * signed compiler will perform arithmetic shift. This last operation
+	 * is implementation defined in C but defined in the compiler's
+	 * documentation and common for modern compilers and CPUs.
+	 */
+	return ((signed int)((unsigned int)x << 20)) >> 20;
+}
+
+static int spi_cs_ctrl(struct paw3212_data *dev_data, bool enable)
+{
+	u32_t val = (enable) ? (0) : (1);
+	int err;
+
+	if (!enable) {
+		k_busy_wait(T_NCS_SCLK);
+	}
+
+	err = gpio_pin_write(dev_data->cs_gpio_dev, CONFIG_PAW3212_CS_GPIO_PIN,
+			     val);
+	if (err) {
+		LOG_ERR("SPI CS ctrl failed");
+	}
+
+	if (enable) {
+		k_busy_wait(T_NCS_SCLK);
+	}
+
+	return err;
+}
+
+static int reg_read(struct paw3212_data *dev_data, u8_t reg, u8_t *buf)
+{
+	int err;
+
+	__ASSERT_NO_MSG((reg & SPI_WRITE_BIT) == 0);
+
+	err = spi_cs_ctrl(dev_data, true);
+	if (err) {
+		return err;
+	}
+
+	/* Write register address. */
+	const struct spi_buf tx_buf = {
+		.buf = &reg,
+		.len = 1
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1
+	};
+
+	err = spi_write(dev_data->spi_dev, &spi_cfg, &tx);
+	if (err) {
+		LOG_ERR("Reg read failed on SPI write");
+		return err;
+	}
+
+	k_busy_wait(T_SRAD);
+
+	/* Read register value. */
+	struct spi_buf rx_buf = {
+		.buf = buf,
+		.len = 1,
+	};
+	const struct spi_buf_set rx = {
+		.buffers = &rx_buf,
+		.count = 1,
+	};
+
+	err = spi_read(dev_data->spi_dev, &spi_cfg, &rx);
+	if (err) {
+		LOG_ERR("Reg read failed on SPI read");
+		return err;
+	}
+
+	err = spi_cs_ctrl(dev_data, false);
+	if (err) {
+		return err;
+	}
+
+	k_busy_wait(T_SRX);
+
+	return 0;
+}
+
+static int reg_write(struct paw3212_data *dev_data, u8_t reg, u8_t val)
+{
+	int err;
+
+	__ASSERT_NO_MSG((reg & SPI_WRITE_BIT) == 0);
+
+	err = spi_cs_ctrl(dev_data, true);
+	if (err) {
+		return err;
+	}
+
+	u8_t buf[] = {
+		SPI_WRITE_BIT | reg,
+		val
+	};
+	const struct spi_buf tx_buf = {
+		.buf = buf,
+		.len = ARRAY_SIZE(buf)
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1
+	};
+
+	err = spi_write(dev_data->spi_dev, &spi_cfg, &tx);
+	if (err) {
+		LOG_ERR("Reg write failed on SPI write");
+		return err;
+	}
+
+	k_busy_wait(T_SCLK_NCS_WR);
+
+	err = spi_cs_ctrl(dev_data, false);
+	if (err) {
+		return err;
+	}
+
+	k_busy_wait(T_SWX);
+
+	return 0;
+}
+
+static void irq_handler(struct device *gpiob, struct gpio_callback *cb,
+			u32_t pins)
+{
+	sensor_trigger_handler_t handler;
+
+	k_spinlock_key_t key = k_spin_lock(&paw3212_data.lock);
+	handler = paw3212_data.data_ready_handler;
+	k_spin_unlock(&paw3212_data.lock, key);
+
+	struct sensor_trigger trig = {
+		.type = SENSOR_TRIG_DATA_READY,
+		.chan = SENSOR_CHAN_ALL,
+	};
+
+	handler(DEVICE_GET(paw3212), &trig);
```

Comment:
Looks like the handler cannot be called directly from here, see [sensor.h](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/sensor.h#L360). You should use the system work queue or a dedicated thread for it.

Reply:
I have overseen it... this is not something I like as it will impact the performance :(
We need to fix another sensor and qdec too.

Reply:
Fixed